store/primary/cache.go

```diff
@@ -0,0 +1,312 @@
+			return nil, false
+		}
+
+		// Search previous if not found in current
+		v, found = c.previous.Get(k)
+		if !found {
+			return nil, false
+		}
+
+		// Pull the interned entries for these values forward from the previous
+		// cache to keep using the same pointers as the va used in the cache
+		// values that get pulled forward.
+		values := v.([]*store.IndexEntry)
+		for i, val := range values {
+			values[i] = c.internEntry(val)
+		}
+
+		// Move the value found in the previous tree into the current one.
+		c.current.Put(k, values)
+		c.previous.Delete(k)
+	}
+	return v.([]*store.IndexEntry), true
+}
+
+// put stores an entry in the cache if the entry is not already stored.
+// Returns true if a new entry was added to the cache.
+func (c *syncCache) put(k string, entry store.IndexEntry, rotateSize int) bool {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	// Get from current or previous cache
+	old, found := c.getNoLock(k)
+	// If found values(s) then check the value to put is already there.
+	if found && duplicateEntry(entry, old) {
+		return false
+	}
+
+	if c.current.Len() >= rotateSize {
+		c.rotate()
+	}
+
+	c.current.Put(k, append(old, c.internEntry(&entry)))
+	return true
+}
+
+func (c *syncCache) putInterned(k string, ent *store.IndexEntry, rotateSize int) bool {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	old, found := c.getNoLock(k)
+	if found && duplicateEntry(*ent, old) {
+		return false
+	}
+
+	if c.current.Len() >= rotateSize {
+		c.rotate()
+	}
+
+	c.current.Put(k, append(old, ent))
+	return true
+}
+
+func (c *syncCache) putMany(keys []string, entry store.IndexEntry, rotateSize int) int {
+	var count int
+	var ent *store.IndexEntry
+
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	for _, k := range keys {
+		old, found := c.getNoLock(k)
+		if found && duplicateEntry(entry, old) {
+			continue
+		}
+		if ent == nil {
+			ent = c.internEntry(&entry)
+		}
+		c.current.Put(k, append(old, ent))
+		count++
+	}
+
+	if c.current.Len() > rotateSize {
+		c.rotate()
+	}
+
+	return count
+}
+
+func (c *syncCache) remove(k string, entry store.IndexEntry) bool {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	removed := removeEntry(c.current, k, entry)
+	if c.previous != nil && removeEntry(c.previous, k, entry) {
+		removed = true
+	}
+	return removed
+}
+
+func (c *syncCache) removeProvider(providerID peer.ID) int {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	count := removeProviderEntries(c.current, providerID)
+	removeProviderInterns(c.curEnts, providerID)
+	if c.previous != nil {
+		count += removeProviderEntries(c.previous, providerID)
+		if c.prevEnts != nil {
+			removeProviderInterns(c.prevEnts, providerID)
+		}
+	}
+	return count
+}
+
+func (c *syncCache) rotate() {
+	c.previous, c.current = c.current, radixtree.New()
+	c.prevEnts, c.curEnts = c.curEnts, radixtree.New()
+	c.rotations++
+}
+
+func (c *syncCache) internEntry(entry *store.IndexEntry) *store.IndexEntry {
+	k := string(entry.ProvID) + cidToKey(entry.PieceID)
```

Comment:
We may need to change this in the future. If we end up allowing providers to add any kind of metadata related to the CID, we need an alternative way of identifying the internEntry key, as a pieceIDs may not be provided.

Reply:
we should expect to continue having strict constraints on the amount and structure of metadata, since such data has relatively high cost in its need to be replicated for content routing. We should expect large number of cids in an update all of which are allowed to point to one 'record' of the provider issuing them, and ideally metadata that is opaque to the indexing subsystem but which serializes to, say, less than 100 bytes.

Reply:
I will change the implementation in a follow-up PR